### DISCUSSION TITLE
exec plugin with successCmd can causes *first release only* of maintenance branch to fail

### DISCUSSION THREAD

**Discussion by djeanprost** [semantic-release/semantic-release#3523] [Q&A] [answered]
Hello
Don't know if it's the right place to raise this issue, but let's hope.

I have a problem when using `@semantic-release/exec` is used in my configuration when releasing a maintenance version. In such case, the commit analyzer is not run, and my pipeline breaks :
Semantic configuration :
```javascript
module.exports = {
  branches: [
    '+([0-9])?(.{+([0-9]),x}).x',
    'release',
  ],
  plugins: [
    "@semantic-release/commit-analyzer",
    "@semantic-release/release-notes-generator",
    [
      "@terrestris/maven-semantic-release",
      {
        "settingsPath": process.env.MAVEN_SETTINGS,
        "clean": false,
        "updateSnapshotVersion": "true",
        "processAllModules": "true"
      }
    ],
    [
      "@semantic-release/git",
      {
        assets: ["**/pom.xml"]
      },
    ],
    "@semantic-release/gitlab",
    ["@semantic-release/exec", {
      "publishCmd": "echo VERSION_BUILD=${nextRelease.version} > version.env",
      "successCmd": "cat version.env && touch RELEASE_DONE"
    }]
  ],
};
```

```
├── @semantic-release/commit-analyzer@13.0.0
├── @semantic-release/exec@6.0.3
├── @semantic-release/gitlab@13.2.1
├── @semantic-release/release-notes-generator@14.0.1
├── @terrestris/maven-semantic-release@3.1.0
├── conventional-changelog-conventionalcommits@8.0.0
└── semantic-release@24.2.0
```

which gives this execution log : 
```
[semantic-release] › ℹ  Running semantic-release version 24.2.0
[semantic-release] › ✔  Loaded plugin "verifyConditions" from "@terrestris/maven-semantic-release"
[semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/git"
[semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/gitlab"
[semantic-release] › ✔  Loaded plugin "verifyConditions" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"
[semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "verifyRelease" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/release-notes-generator"
[semantic-release] › ✔  Loaded plugin "generateNotes" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "prepare" from "@terrestris/maven-semantic-release"
[semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/git"
[semantic-release] › ✔  Loaded plugin "prepare" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "publish" from "@terrestris/maven-semantic-release"
[semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/gitlab"
[semantic-release] › ✔  Loaded plugin "publish" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "addChannel" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "success" from "@terrestris/maven-semantic-release"
[semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/gitlab"
[semantic-release] › ✔  Loaded plugin "success" from "@semantic-release/exec"
[semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/gitlab"
[semantic-release] › ✔  Loaded plugin "fail" from "@semantic-release/exec"
[semantic-release] › ✔  Run automated release from branch 2.2.x on repository https://gitlab-ci-token:[secure]@git.toto.fr/toto/titi.git
[semantic-release] › ✔  Allowed to push to the Git repository
[semantic-release] › ℹ  Start step "verifyConditions" of plugin "@terrestris/maven-semantic-release"
[semantic-release] [@terrestris/maven-semantic-release] › ℹ  Testing if mvn exists
[semantic-release] › ✔  Completed step "verifyConditions" of plugin "@terrestris/maven-semantic-release"
[semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/git"
[semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/git"
[semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/gitlab"
[semantic-release] [@semantic-release/gitlab] › ℹ  Verify GitLab authentication (https://git.toto.fr/api/v4)
[semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/gitlab"
[semantic-release] › ℹ  Start step "verifyConditions" of plugin "@semantic-release/exec"
[semantic-release] › ✔  Completed step "verifyConditions" of plugin "@semantic-release/exec"
[semantic-release] › ℹ  Found 2 commits since last release
[semantic-release] › ℹ  Start step "generateNotes" of plugin "@semantic-release/release-notes-generator"
[semantic-release] › ✔  Completed step "generateNotes" of plugin "@semantic-release/release-notes-generator"
[semantic-release] › ℹ  Start step "generateNotes" of plugin "@semantic-release/exec"
[semantic-release] › ✔  Completed step "generateNotes" of plugin "@semantic-release/exec"
[semantic-release] › ✔  Add channel 2.2.x to tag v2.2.0
[semantic-release] › ℹ  Start step "addChannel" of plugin "@semantic-release/exec"
[semantic-release] › ✔  Completed step "addChannel" of plugin "@semantic-release/exec"
[semantic-release] › ℹ  Start step "success" of plugin "@terrestris/maven-semantic-release"
[semantic-release] [@terrestris/maven-semantic-release] › ℹ  Update pom.xml to next snapshot version
[semantic-release] [@terrestris/maven-semantic-release] › ℹ  Staging all changed files: pom.xml, referentiel-rest-controller/pom.xml, referentiel-persistence/pom.xml, referentiel-integration/pom.xml, referentiel-domain/pom.xml, referentiel-boot/pom.xml, referentiel-api-specification/pom.xml, referentiel-api-dto/pom.xml, referentiel-integration/target/generated-sources/openapi/pom.xml
[semantic-release] [@terrestris/maven-semantic-release] › ℹ  Committing all changed pom.xml
[semantic-release] [@terrestris/maven-semantic-release] › ℹ  Pushing commit
[semantic-release] › ✔  Completed step "success" of plugin "@terrestris/maven-semantic-release"
[semantic-release] › ℹ  Start step "success" of plugin "@semantic-release/gitlab"
[semantic-release] › ✔  Completed step "success" of plugin "@semantic-release/gitlab"
[semantic-release] › ℹ  Start step "success" of plugin "@semantic-release/exec"
[semantic-release] [@semantic-release/exec] › ℹ  Call script cat version.env && touch RELEASE_DONE
[semantic-release] › ✘  Failed step "success" of plugin "@semantic-release/exec"
[semantic-release] › ✘  An error occurred while running semantic-release: Error: Command failed with exit code 1: cat version.env && touch RELEASE_DONE
```

in this execution log, you can see that there are two lines
```
[semantic-release] › ℹ  Found 2 commits since last release
[semantic-release] › ℹ  Start step "generateNotes" of plugin "@semantic-release/release-notes-generator"
```
and there is no commit analyzer ran at all.

Complete log is here [execution.log](https://github.com/user-attachments/files/17922797/execution.log)

Can some tell me what's wrong in my settings there please ?
Please note that the error at the end is from my point of view a consequence of the fact that the whole process didn't run correctly, the root cause being IMO the lack of commit analyzer.

**Comment by djeanprost** [comment]
@travi I'm really sorry to pinging you so rudely, but can you have a look at my question which really blocks me please 🙏
I really don't understand what happens there, and it prevents my team to release any maintenance branch.
I really thank you in advance

**Comment by travi** [comment]
based on a quick look, i dont see anything obvious. order of defining the plugins is important in your config, but it looks like that is correct already.

this suggests that it should be ready to be used:

> [semantic-release] › ✔  Loaded plugin "analyzeCommits" from "@semantic-release/commit-analyzer"

if you arent already, you could try to use the [debug flag](https://semantic-release.gitbook.io/semantic-release/usage/configuration#debug) when you execute. you may need to create a minimal reproduction in a public project for us to dig further from this side.

**Reply by djeanprost** [reply]
Hello @travi Thank you for taking time to help me.
Debug flag was added, and the log is available in my initial message.
Do you need more informations from me to test ? With or without exec plugin ?
I can tell you that the problem **doesn't occur on **main branch**** when doing release, **only on maintenance branch**. It seems to me that **adding the channel** to the branch is related to the stuff ? What do you think ?

**Reply by djeanprost** [reply]
forgot to give you plugins versions
```
├── @semantic-release/commit-analyzer@13.0.0
├── @semantic-release/exec@6.0.3
├── @semantic-release/gitlab@13.2.1
├── @semantic-release/release-notes-generator@14.0.1
├── @terrestris/maven-semantic-release@3.1.0
├── conventional-changelog-conventionalcommits@8.0.0
└── semantic-release@24.2.0
```

**Reply by travi** [reply]
You don't need to install commit analyzer or release notes generator directly since they are already dependencies of semantic-release. 

> I can tell you that the problem doesn't occur on main branch when doing release, only on maintenance branch. It seems to me that adding the channel to the branch is related to the stuff ?

You have to many variables. You need to start by narrowing down the situation and understanding which variable is causing the behavior difference. Adding more variables makes it even harder to suggest what the true issue might be. 

> Do you need more informations from me to test ?

As mentioned before, without a minimal public reproduction, there is not much I can do beyond what is not obvious from the logs. By creating a public reproduction, you can also try to reduce the problem and maybe even identify the problem yourself. I want to help you, but you have to help me do so. My time available to provide support is limited, so the more you can help me optimize that to focus on the true problem, the more helpful I can be

**Reply by djeanprost** [reply]
Hello @travi 
With a bit of work, I finally managed to reproduce my case in this public repo https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533553074
this job execution does show that `commit-analyzer` is not run at all in this case.

BUT, if I run it a second time after the failure, then the [second time](https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533615489) it will work surprisingly. From my understanding it has to do with the fact that during second execution, the channel stuff is not exactly the same than during first one.

Another test for a branch where exec plugin is not used : it works at the 1st time : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533673385

So if you have any idea with that, I take it.
If you needed write access to the repository to test things, I can give you permission.

**Reply by djeanprost** [reply]
@pvdlg ? sorry for getting you there, but maybe you can help with this bug-a-like ?

**Comment by travi** [comment]
> when releasing a maintenance version

when you are releasing a maintenance version, does a release for the range defined for the branch already exist or are you creating the maintenance branch with the expectation that the first release for that range will be created? i havent had the time to completely confirm my suspicion, but it looks like you are approaching this workflow expecting the latter

**Reply by djeanprost** [reply]
Hello @travi I'm not sure I completely understood your question, but here is my workflow just in case it's your question :
- create a couple of "main releases" : v1.1.0, v1.2.0, v1.3.0...
- bug is discovered on 1.2.0, I need to fix and get a release
- So I create a branch 1.2.x from tag v1.2.0
- add a `fix:` commit on branch 1.2.x, and expect to get a new release called v1.2.1 on this maintenance branch.

Does it answer your question ?
With this workflow, if I use exec plugin it doesn't work as I explained. Without, it does.

**Reply by travi** [reply]
yes, that answers my question. we've had several situtions lately where folks misunderstan the intend of the maintenance workflow, so i wanted to make sure similar confusion didnt exist here. i'll need to find time to trace through this more closely and walk through the implementation. i can't promise a timeline, so you are welcome to look through the details if you are able to dig deeper before i can.

**Reply by djeanprost** [reply]
again, the problem only occurs if exec plugin is used or not, as mentionned in https://github.com/semantic-release/semantic-release/discussions/3523#discussioncomment-11450807

**Reply by travi** [reply]
what is the purpose of the release branch in your workflow and what is its relationship to the default branch?

**Reply by djeanprost** [reply]
Hello @travi 
I've seen that I forgot to answer your question.
My default branch which is `develop` in my workflow is where I add commits (`fix` or `feat`) in order to accumulate new features for next release.
Then when I'm satisfied with the content, I create a MR from this `develop` branch to `release` branch in order to launch the release process with new commits.
Does this answer help you to understand ?

**Reply by travi** [reply]
> My default branch which is `develop` in my workflow is where I add commits (`fix` or `feat`) in order to accumulate new features for next release.
Then when I'm satisfied with the content, I create a MR from this develop branch to release branch in order to launch the release process with new commits.

what you are describing is gitflow, which is a workflow that is not officially supported by semantic-release. while semantic-release is configurable enough to simulate gitflow, we do not recommend it and do not provide support for such configuration. it is a "at your own risk" situation if you choose to follow gitflow with semantic-release

> need to create a minimal reproduction in a public project for us to dig further from this side

i think we are back to this request. while making a public reproduction helped me understand your intended workflow better, it skipped over the "minimal" part of the request. in order to dig deeper into this situation, we need to reproduce this in the context of the supported workflows. deviating from the supported workflows introduces variables beyond your theory that the exec plugin is the cause. that means we need a configuration where the primary release branch is the default branch and all features are merged to the default branch rather than through a develop branch first. this can help us eliminate the unsupported workflow from the equation and determine if your theory of being related to the exec plugin is correct

**Reply by djeanprost** [reply]
Hello @travi 
Fortunatly I managed to reproduce the problem in a way that should suits your requirement to stick to initial semantic release way of thinking.
I changed the settings to this one where `main` is the default branch in my repo. I also changed my pipeline definition accordingly to allow semantic job to run on this branch.
```javascript
module.exports = {
    branches: [
        '+([0-9])?(.{+([0-9]),x}).x',
        'main',
    ],
    plugins: [
        "@semantic-release/commit-analyzer",
        "@semantic-release/release-notes-generator",
        [
        "@terrestris/maven-semantic-release",
        {
            "clean": false,
            "mavenTarget": "package jib:build",
            "updateSnapshotVersion": false,
            "processAllModules": true
        }
        ],
        [
        "@semantic-release/git",
        {
            assets: ["**/pom.xml"]
        },
        ],
        "@semantic-release/gitlab",
        ["@semantic-release/exec", {
        "publishCmd": "echo VERSION_BUILD=${nextRelease.version} > version.env",
        "successCmd": "cat version.env && touch RELEASE_DONE"
        }]
    ],
};
```

```yaml
release-job:
  stage: release
  rules:
    # This branch definition has to be the same than the branch configured in semantic release
    - if: $CI_COMMIT_REF_NAME == "main" || $CI_COMMIT_REF_NAME =~ /^[0-9]+\.[0-9]+\.x$/
  image: $MAVEN_IMAGE_NAME
  variables:
    NODE_OPTIONS: "--disable-warning=ExperimentalWarning"
    DEBUG: "semantic-release:*"
  before_script:
    - curl -fsSL https://deb.nodesource.com/setup_23.x -o nodesource_setup.sh
    - bash nodesource_setup.sh
    - apt-get install -y nodejs
    - node -v
    - echo "Node is installed"
  script:
    - npm install semantic-release@24
    - npm install @semantic-release/gitlab@13
    - npm install @terrestris/maven-semantic-release@3
    - npm install @semantic-release/exec@6
    - npm install conventional-changelog-conventionalcommits@8
    - npx semantic-release --extends ./semantic-release-config.js
```

And this will give the same results I've been struggling with :
- execution on **main** branch --> gives the expected result : new minor version is produced
   - Example of https://gitlab.com/djeanprost/semantic-release-issue/-/releases/v1.4.0, with associated tag and job execution : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8788498271
- execution on **maintenance** branch fails :
   - First execution fails as always : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8788581668 : `commit-analyzer` still not executed on first run
   - Second execution is in success : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8788614249 : `commit-analyzer` is executed this time https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8788614249#L371
   - First execution does work if exec plugin is not used : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8793581842#L354

--> Back from the start

I hope this will help you to drill down on what is the cause of the problem, and by the way I really thank you for helping. The lead that seems the most interesting to me even if I'm limited in my understanding of it is @kilianpaquier question : https://github.com/semantic-release/semantic-release/discussions/3523#discussioncomment-11524713

Looking forward reading from you 😄

**Reply by djeanprost** [reply]
@travi any updates on your side maybe 😁 ?

**Reply by djeanprost** [reply]
@travi sorry for being so rude, but this issue is still with no solution 😐. From my point of view, I reproduced everything you requested. If there are still missing feedback from me, please don't hesitate to ask

**Reply by travi** [reply]
i'm sorry that i don't have a better answer for you. we appreciate the effort you've already put into reproducing this problem. i've investigated far enough to believe that it will take quite a bit of effort to really get to the bottom of this. unfortunately, i have not had the time lately to dedicate that level of focus to this. if getting past this is important to you, you are likely to get to an answer by investigating more deeply yourself and sharing your findings. in the end, we will need to reproduce whatever problem is causing this as a test, so as you investigate keep that in mind in case it can help guide you to the steps we'll need to get to for this to get resolved.

**Comment by kilianpaquier** [comment]
Hi !

Could it come from the fact that for this bloc of semantic-release execution, only `generateNotes`, `addChannel` and `success` steps are executed ?

https://github.com/semantic-release/semantic-release/blob/6022328c6ba83cb2ae587d0d88e8c266de858648/index.js#L110-L153

However, at first glance, I don't see why it would work the second time ...

What makes me say it could come from here (may be an expected behavior) is that on @djeanprost first CI error, this line appears [Found 4 commits since last release](https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533553074#L354) and then it goes straight to `generateNotes`.

From the look of it, the only execution succession giving this is [there](https://github.com/semantic-release/semantic-release/blob/6022328c6ba83cb2ae587d0d88e8c266de858648/index.js#L120-L121).

And it does seem that this execution flow is specific to maintenance branches ? https://github.com/semantic-release/semantic-release/blob/6022328c6ba83cb2ae587d0d88e8c266de858648/index.js#L117-L121

**Reply by djeanprost** [reply]
> However, at first glance, I don't see why it would work the second time ...

the only difference I can see in 1st execution and in 2nd one is that on 2nd one, channel 1.2.x is already present on the tag/branch. I guess it was added on first time.

**Comment by travi** [comment]
in attempting to look into this issue, i've been looking through historical discussions to see if others have come across any similar behaviors. 

while doing so, i spotted a [previous thread that you started](https://github.com/semantic-release/semantic-release/discussions/2990) where you mentioned behavior that wasnt working correctly with the first execution, but was working with a second run. i see that i didnt manage to respond to that one, so apologies for that. did you end up solving that situation? 

it is interesting that you've seen misbehavior on first run but correct behavior on a second run in both cases. this makes me further wonder if there is something about your workflow that i havent grasped yet that might be presenting this situation. you havent described anything yet that gives me any real confidence that this is the case, but this is at least an interesting pattern that stands out to me

**Reply by djeanprost** [reply]
Hello @travi
Yes I previously opened [this discussion](https://github.com/semantic-release/semantic-release/discussions/2990) a couple of weeks ago, and didn't get any clue on what happened. Reopened this one which is a duplicate of previous one because I've managed to understand a bit better of the execution context, and I'm now sure that the presence of exec plugin in the plugins list does change something in the execution flow.

Indeed there could be something in my branch workflow or release management, I agree. But what I cannot understand is that depending on the presence of exec plugin, it works well or not.
Let me summarize again : 
- without exec plugin in the flow : it works at first run on a maintenance branch : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533673385
- with exec plugin in the flow : 
  - first run doesn't work (because commits are not analyzed ?) : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533553074
  - second run does work : https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533615489

The only difference I can spot in 1st and 2nd run with exec plugin is that at first time, channel is not set on the maintenance branch, and during 2nd one, it's set (I guess by previous run), although it's not present too on single run when exec plugin is not used.

Do you need write access to my personnal repo to test on your side or reproduce something ? I can give you permission if you request it.
We didn't get any answer yet, but I thank you for the time you spend on this issue.

**Reply by djeanprost** [reply]
@travi  Did you have a look at what is mentionned here ? https://github.com/semantic-release/semantic-release/discussions/3523#discussioncomment-11524713

At least from what is said, the piece of code that is highlighted could explain something ?

**Comment by babblebey** [comment]
Hey @djeanprost,

After looking at this and the threads in this discussion, I think there's an imperative part of the error that you might be unintentionally leaving it out.... See below..

```
[4:42:21 PM] [semantic-release] [@semantic-release/exec] › ℹ  Call script cat version.env && touch RELEASE_DONE
cat: version.env: No such file or directory
[4:42:21 PM] [semantic-release] › ✘  Failed step "success" of plugin "@semantic-release/exec"
[4:42:21 PM] [semantic-release] › ✘  An error occurred while running semantic-release: Error: Command failed with exit code 1: cat version.env && touch RELEASE_DONE
cat: version.env: No such file or directory
    at makeError (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/node_modules/execa/lib/error.js:60:11)
    at handlePromise (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/node_modules/execa/index.js:118:26)
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async module.exports (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/lib/exec.js:16:11)
    at async success (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/index.js:107:5)
    at async validator (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/normalize.js:36:24)
    at async file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:38:36
    at async file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:32:5
    at async pluginsConfigAccumulator.<computed> [as success] (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/index.js:87:11)
    at async run (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/index.js:151:7) {
  shortMessage: 'Command failed with exit code 1: cat version.env && touch RELEASE_DONE',
  command: 'cat version.env && touch RELEASE_DONE',
  escapedCommand: '"cat version.env && touch RELEASE_DONE"',
  exitCode: 1,
  signal: undefined,
  signalDescription: undefined,
  stdout: '',
  stderr: 'cat: version.env: No such file or directory',
  failed: true,
  timedOut: [secure],
  isCanceled: [secure],
  killed: [secure],
  pluginName: '@semantic-release/exec'
}
AggregateError: 
    Error: Command failed with exit code 1: cat version.env && touch RELEASE_DONE
    cat: version.env: No such file or directory
        at makeError (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/node_modules/execa/lib/error.js:60:11)
        at handlePromise (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/node_modules/execa/index.js:118:26)
        at async module.exports (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/lib/exec.js:16:11)
        at async success (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/index.js:107:5)
        at async validator (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/normalize.js:36:24)
        at async file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:38:36
        at async file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:32:5
        at async pluginsConfigAccumulator.<computed> [as success] (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/index.js:87:11)
        at async run (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/index.js:151:7)
    at file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:55:13
    at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
    at async pluginsConfigAccumulator.<computed> [as success] (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/index.js:87:11)
    at async run (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/index.js:151:7)
    at async Module.default (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/index.js:278:22)
    at async default (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/cli.js:55:5) {
  errors: [
    Error: Command failed with exit code 1: cat version.env && touch RELEASE_DONE
    cat: version.env: No such file or directory
        at makeError (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/node_modules/execa/lib/error.js:60:11)
        at handlePromise (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/node_modules/execa/index.js:118:26)
        at process.processTicksAndRejections (node:internal/process/task_queues:105:5)
        at async module.exports (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/lib/exec.js:16:11)
        at async success (/builds/djeanprost/semantic-release-issue/node_modules/@semantic-release/exec/index.js:107:5)
        at async validator (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/normalize.js:36:24)
        at async file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:38:36
        at async file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/pipeline.js:32:5
        at async pluginsConfigAccumulator.<computed> [as success] (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/lib/plugins/index.js:87:11)
        at async run (file:///builds/djeanprost/semantic-release-issue/node_modules/semantic-release/index.js:151:7) {
      shortMessage: 'Command failed with exit code 1: cat version.env && touch RELEASE_DONE',
      command: 'cat version.env && touch RELEASE_DONE',
      escapedCommand: '"cat version.env && touch RELEASE_DONE"',
      exitCode: 1,
      signal: undefined,
      signalDescription: undefined,
      stdout: '',
      stderr: 'cat: version.env: No such file or directory',
      failed: true,
      timedOut: [secure],
      isCanceled: [secure],
      killed: [secure],
      pluginName: '@semantic-release/exec'
    }
  ]
}
```

It point out how the `version.env` file doesn't exist for it to work with in the exec's `success` step of the release. From your config you stated above it appears that you intend for this file to be created in the exec's `publish` step....

BUT the WEIRD BUT is that the `publish` step is missing the logs at https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/8533553074; this can in-turn mean that the `publish` step didn't run at all (the `publish` step in a release should run before the `success` step). this is the step where your `version.env` file is meant to be created.

 Why? How come the `publish` step is skipped is the thing to find out.

**Reply by djeanprost** [reply]
Hello @babblebey 
Thank you for joining this incredible discussion.
Yes indeed the error you mention (lack of version.env) is the consequence (to my mind) of the original problem which is the fact, for what we are in the diagnosis process, is the fact that the commit-analyzer is not run in my case.
For where we are now, what should be investigate is @kilianpaquier message which seems the most promising from my point of view : https://github.com/semantic-release/semantic-release/discussions/3523#discussioncomment-11524713. Could you investigate his suggestion please maybe ?

**Comment by kilianpaquier** [comment, marked as answer]
Hello @travi,

After further investigation, it's really coming from here: 

https://github.com/semantic-release/semantic-release/blob/master/index.js#L151

### What is happening:
- A release was made some time ago giving the tag, let's say `1.3.0`
- Release were made after that, giving (at least) one new minor version, `1.4.0`, `1.5.0`, etc.
- A bugfix must be made on `1.3.0` for any justified reason (security, etc.)
- I (a random developer), create the maintenance branch `1.3.x` and push (directly or with a pull request it doesn't matter) the appropriate fix on it
- A CI then run, let's concentrate on **semantic-release** job, keep in mind the released version should be `1.3.1`
- During **semantic-release** job, the execution retrieves branches (and associated channels), however, channel `1.3.x` doesn't exist and isn't associated to tag `1.3.0`
- Execution enters https://github.com/semantic-release/semantic-release/blob/master/index.js#L112 to add the appropriate **git** notes to tag `1.3.0` (to associate channel). Which by the way didn't know this feature existed before now !
- It adds and pushes **git** notes and then, runs `plugins.success` step

### Why does it fail on @djeanprost specific case ?

Simple, **`plugins.publish`**, in charge in his case to `echo VERSION=${nextRelease.version} > version.env` **isn't executed** (not before https://github.com/semantic-release/semantic-release/blob/master/index.js#L215). As such, `cat version.env` (a missing file) ends with an exit code 1.

### Multiple questions come to mind with this specific discussion

Historically, from oldest to newest (yes I checked previous commits impacting this code), it seems tags were created, which was a justified reason to call `plugins.success`. However, code evolved to only annotate tags right now. 

Which poses with, do we still need to run `plugins.success` since there's no more release made here ? And in my opinion, there was no release made at all here, never, since neither `plugins.analyzeCommits` nor `plugins.verifyRelease` or `plugins.publish` were (never) executed around this code.

https://github.com/semantic-release/semantic-release/commit/7b4052470b23261c9e679a17bff034da311fd894#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R112-R123

https://github.com/semantic-release/semantic-release/commit/aec96c791f7413dace1bfdca08f7a5cd58cb0f5e#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346R109-R122

https://github.com/semantic-release/semantic-release/commit/b2c1b2c670f8f2dd4da71721ffb329c26e8d2cd7#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L112-R115

Another question that comes to my mind, and I can backup it with this pull request and gitlab run (thanks @djeanprost):

- https://gitlab.com/djeanprost/semantic-release-issue/-/merge_requests/8. 
- https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/9365017331#L385
- https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/9365017331#L549

It seems on top of that, that this `plugins.success` is executed on the previous version and not the current one being released. Isn't it weird ? As it could notify a success multiple times for a release maybe made days or months ago ?

**Reply by djeanprost** [reply]
As a test, I removed the `cat version.env` part of my `successCMD` of exec plugin, and now it works well as you can see here https://gitlab.com/djeanprost/semantic-release-issue/-/jobs/9364866628
it seems indeed that because of un expected run of `successCmd` in the flow as @kilianpaquier explained, the `cat` was failing thus causing failure of the whole process.
I wrongly associated the presence of `exec plugin` and non execution of `commit-analyzer`, but it was in fact a failing `successCmd` that caused a non execution of `commit-analyzer` and then a failure of the whole process.
As a workaround, I changed my `successCmd` command : https://gitlab.com/djeanprost/semantic-release-issue/-/commit/dba154b901065fd5722953ef1da935ab760e8d7a
The most solid conclusion would be to fix the issue raised by @kilianpaquier.

Thanks everybody for helping.

Changed discussion title to reflect this diagnosis.

**Reply by travi** [reply]
thank you for the further investigation. i will be largely afk for the next couple of weeks before i can attempt to dig deeper to catch up on my understanding.

i think the recommendation i can give in the mean time is that we will need a reproduction in a test to get this resolved. if someone would be willing to add a test to our e2e suite in this project, i think that will be the best place to reproduce this issue fully. while this surfaced using the exec plugin, i think it would be best to simulate the problem without adding a dependency on the actual exec plugin to the project. it should be possible to set up the test using a faked out plugin to accomplish a similar issue. if you run into any problems with the test, feel free to create a draft pr and ask us questions. we will try to help you further as we have availability
